src/blocking/linux.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2022 Stephane Raux & Contributors. Distributed under the 0BSD license.
+
+//! Platform-specific implementation for Linux.
+
+use crate::Error;
+use std::path::PathBuf;
+use std::{fs, io};
+
+pub(crate) const BACKLIGHT_DIR: &str = "/sys/class/backlight";
+
+#[derive(Debug)]
+pub(crate) struct BlockingDeviceImpl {
+    device: String,
+}
+
+impl crate::blocking::Brightness for BlockingDeviceImpl {
+    fn device_name(&self) -> Result<String, Error> {
+        Ok(self.device.clone())
+    }
+
+    fn get(&self) -> Result<u32, Error> {
+        let max = read_value(&self.device, Value::Max)?;
+        let actual = read_value(&self.device, Value::Actual)?;
+        let percentage = if max == 0 {
+            0
+        } else {
+            (actual * 100 / max) as u32
+        };
+        Ok(percentage)
+    }
+
+    fn set(&self, percentage: u32) -> Result<(), Error> {
+        let percentage = percentage.min(100);
+        let max = read_value(&self.device, Value::Max)?;
+        let desired_value = (u64::from(percentage) * u64::from(max) / 100) as u32;
+        let desired = ("backlight", &self.device, desired_value);
+        let bus =
+            zbus::blocking::Connection::system().map_err(|e| Error::SettingBrightnessFailed {
+                device: self.device.clone(),
+                source: e.into(),
+            })?;
+        let response = bus.call_method(
+            Some("org.freedesktop.login1"),
+            "/org/freedesktop/login1/session/auto",
+            Some("org.freedesktop.login1.Session"),
+            "SetBrightness",
+            &desired,
+        );
+        match response {
+            Ok(_) => Ok(()),
+            Err(zbus::Error::MethodError(..)) => {
+                // Setting brightness through dbus may not work on older systems that don't have
+                // the `SetBrightness` method. Fall back to writing to the brightness file (which
+                // requires permission).
+                set_value(&self.device, desired_value)?;
+                Ok(())
+            }
+            Err(e) => Err(Error::SettingBrightnessFailed {
+                device: self.device.clone(),
+                source: e.into(),
+            }),
+        }
+    }
+}
+
+pub(crate) fn brightness_devices() -> Result<Vec<BlockingDeviceImpl>, SysError> {
```

Comment:
ok - I've done this so that the public API returns `impl Iterator`
But I have had to make the internals `Box<dyn Iterator>` because they return different iterators based on the result (iter::once vs iter::map)

Reply:
actually there is an Either in Itertools that works for this :)